File: backends/php-laravel/app/UseCases/Notification/NotificationService.php
<?php

/**
 * This is a program developed by BobTabo.
 *
 * Copyright (c) 2026 BobTabo. All Rights Reserved.
 */

declare(strict_types=1);

namespace App\UseCases\Notification;

use App\Domain\Notification\Condition\NotificationCondition;
use App\Domain\Notification\Entities\Notification;
use App\Domain\Notification\Repositories\NotificationRepository;
use App\Domain\Notification\ValueObjects\NotificationCountsVo;
use App\Domain\Notification\ValueObjects\NotificationListVo;
use App\Domain\Notification\ValueObjects\NotificationSaveVo;
use App\Domain\Staff\Entities\Staff;
use App\Domain\Staff\Repositories\StaffRepository;
use App\Support\Exceptions\AppException;
use App\Support\Mappers\SimpleMapper;
use App\Support\Services\AbstractService;
use App\Support\Traits\EnumValue;
use app\UseCases\Notification\Dtos\NotificationCreateDto;
use App\UseCases\Notification\Dtos\NotificationDto;

/**
 * 通知Serviceクラスです。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 * @package App\UseCases\Notification
 */
class NotificationService extends AbstractService
{
    use EnumValue;

    /**
     * @param NotificationRepository $notificationRepository 通知Repository
     * @param StaffRepository $staffRepository スタッフRepository
     */
    public function __construct(
        private readonly NotificationRepository $notificationRepository,
        private readonly StaffRepository $staffRepository,
    ) {
    }

    /**
     * カーソル付きで通知一覧ページを取得します。
     *
     * @param NotificationDto $dto 通知DTO
     * @return NotificationListVo 通知一覧ValueObject
     */
    public function listPage(NotificationDto $dto): NotificationListVo
    {
        $condition = SimpleMapper::map($dto, NotificationCondition::class);

        $list = $this->notificationRepository->listPage($condition);
        $condition->limit = max(1, min(100, $dto->limit));
        $hasNext = $list->count() > $condition->limit;
        $items = $hasNext ? $list->slice(0, $condition->limit) : $list;

        $nextCursor = null;
        if ($hasNext) {
            $last = $items->last();
            $nextCursor = base64_encode($last->createdAt->format('Y-m-d H:i:s') . ',' . $last->id);
        }

        $values = $list->map(function (Notification $row) use ($dto) {
            return $this->toValues($row->attributesBySnake());
        });

        return new NotificationListVo()->assign([
            'items' => $values->all(),
            'nextCursor' => $nextCursor,
        ]);
    }

    /**
     * 通知件数の集計を取得します。
     *
     * @param NotificationDto $dto 通知DTO
     * @return NotificationCountsVo 通知件数ValueObject
     */
    public function counts(NotificationDto $dto): NotificationCountsVo
    {
        /** @var NotificationCondition $condition */
        $condition = SimpleMapper::map($dto, NotificationCondition::class);
        $total = $this->notificationRepository->counts($condition);

        $condition->countUnread = true;
        $unread = $this->notificationRepository->counts($condition);

        return new NotificationCountsVo()->assign([
            'total' => $total,
            'unread' => $unread,
        ]);
    }

    /**
     * 一括既読などの更新を行います。
     *
     * @param NotificationDto $dto 通知DTO
     * @return NotificationSaveVo 通知更新ValueObject
     */
    public function reads(NotificationDto $dto): NotificationSaveVo
    {
        $condition = SimpleMapper::map($dto, NotificationCondition::class);
        $updated = $this->notificationRepository->updateRead($condition);

        return new NotificationSaveVo()->assign(['updated' => $updated]);
    }

    /**
     * 有効なスタッフ全員へ通知を配信します（ファンアウト）。
     *
     * @param NotificationCreateDto $dto 通知登録DTO
     * @return void
     */
    public function fanOut(NotificationCreateDto $dto): void
    {
        $staffs = $this->staffRepository->findAllActive();

        $notifications = $staffs->map(function (Staff $row) use ($dto) {
            $entity = new Notification();
            $entity->staffId = $row->id;
            $entity->messageType = $dto->messageType;
            $entity->title = $dto->title;
            $entity->message = $dto->message;
            $entity->url = $dto->url;
            $entity->assignCreated($dto->executorId);
            return $this->toValues($entity->attributesBySnake());
        });

        $this->notificationRepository->insertBatch($notifications->all());
    }

    /**
     * 単一通知を部分更新します。
     *
     * @param NotificationDto $dto 通知DTO
     * @return NotificationSaveVo 通知更新ValueObject
     */
    public function read(NotificationDto $dto): NotificationSaveVo
    {
        $condition = SimpleMapper::mapSpecific($dto, NotificationCondition::class, [
            'notificationId' => 'id',
        ]);

        $result = $this->notificationRepository->updateRead($condition);
        if (!$result) {
            throw AppException::notFound('notification_not_found');
        }

        return new NotificationSaveVo()->assign([
            'ok' => true,
            'id' => $dto->notificationId
        ]);
    }
}

File: backends/java-spring/src/main/java/com/authorization/usecases/notification/NotificationService.java
/**
 * This is a program developed by BobTabo.
 *
 * Copyright (c) 2026 BobTabo. All Rights Reserved.
 */
package com.authorization.usecases.notification;

import com.authorization.domain.notification.condition.NotificationCondition;
import com.authorization.domain.notification.entities.Notification;
import com.authorization.domain.notification.mappers.NotificationConditionMapper;
import com.authorization.domain.notification.mappers.NotificationCreateMapper;
import com.authorization.domain.notification.repositories.NotificationRepository;
import com.authorization.domain.notification.valueobjects.NotificationCountsVo;
import com.authorization.domain.notification.valueobjects.NotificationListVo;
import com.authorization.domain.notification.valueobjects.NotificationSaveVo;
import com.authorization.domain.staff.entities.Staff;
import com.authorization.domain.staff.repositories.StaffRepository;
import com.authorization.support.exceptions.AppException;
import com.authorization.support.services.AbstractService;
import com.authorization.usecases.notification.dtos.NotificationCreateDto;
import com.authorization.usecases.notification.dtos.NotificationDto;
import java.util.List;

/**
 * 通知Serviceクラスです。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
public class NotificationService extends AbstractService {

    private final NotificationRepository notificationRepository;
    private final StaffRepository staffRepository;
    private final NotificationConditionMapper conditionMapper;
    private final NotificationCreateMapper createMapper;

    /**
     * コンストラクタ。
     *
     * @param notificationRepository 通知Repository
     * @param staffRepository スタッフRepository
     * @param conditionMapper DTO→Condition マッパー
     * @param createMapper DTO→Entity マッパー
     */
    public NotificationService(
            NotificationRepository notificationRepository,
            StaffRepository staffRepository,
            NotificationConditionMapper conditionMapper,
            NotificationCreateMapper createMapper) {
        this.notificationRepository = notificationRepository;
        this.staffRepository = staffRepository;
        this.conditionMapper = conditionMapper;
        this.createMapper = createMapper;
    }

    /**
     * カーソル付きで通知一覧ページを取得します。
     *
     * @param dto 通知DTO
     * @return 通知一覧ValueObject
     */
    public NotificationListVo listPage(NotificationDto dto) {
        NotificationCondition condition = conditionMapper.toCondition(dto);
        condition.setLimit(Math.max(1, Math.min(100, dto.getLimit())));

        List<Notification> list = notificationRepository.listPage(condition);
        boolean hasNext = list.size() > condition.getLimit();
        List<Notification> items = hasNext ? list.subList(0, condition.getLimit()) : list;

        String nextCursor = null;
        if (hasNext) {
            Notification last = items.get(items.size() - 1);
            String raw = last.getCreatedAt() + "," + last.getId();
            nextCursor = java.util.Base64.getEncoder().encodeToString(raw.getBytes());
        }

        NotificationListVo vo = new NotificationListVo();
        vo.setItems(items);
        vo.setNextCursor(nextCursor);
        return vo;
    }

    /**
     * 通知件数の集計を取得します。
     *
     * @param dto 通知DTO
     * @return 通知件数ValueObject
     */
    public NotificationCountsVo counts(NotificationDto dto) {
        NotificationCondition condition = conditionMapper.toCondition(dto);
        int total = notificationRepository.counts(condition);

        condition.setCountUnread(true);
        int unread = notificationRepository.counts(condition);

        NotificationCountsVo vo = new NotificationCountsVo();
        vo.setTotal(total);
        vo.setUnread(unread);
        return vo;
    }

    /**
     * 一括既読などの更新を行います。
     *
     * @param dto 通知DTO
     * @return 通知更新ValueObject
     */
    public NotificationSaveVo reads(NotificationDto dto) {
        NotificationCondition condition = conditionMapper.toCondition(dto);
        int updated = notificationRepository.updateRead(condition);

        NotificationSaveVo vo = new NotificationSaveVo();
        vo.setUpdated(updated);
        return vo;
    }

    /**
     * 有効なスタッフ全員へ通知を配信します（ファンアウト）。
     *
     * @param dto 通知登録DTO
     */
    public void fanOut(NotificationCreateDto dto) {
        List<Staff> staffs = staffRepository.findAllActive();

        List<Notification> notifications = staffs.stream().map(staff -> {
            Notification entity = createMapper.toEntity(dto);
            entity.setStaffId(staff.getId());
            entity.assignCreated(dto.getExecutorId());
            return entity;
        }).toList();

        notificationRepository.insertBatch(notifications);
    }

    /**
     * 単一通知を部分更新します。
     *
     * @param dto 通知DTO
     * @return 通知更新ValueObject
     */
    public NotificationSaveVo read(NotificationDto dto) {
        NotificationCondition condition = conditionMapper.toCondition(dto);
        condition.setId(dto.getNotificationId());

        int result = notificationRepository.updateRead(condition);
        if (result == 0) {
            throw AppException.notFound("notification_not_found");
        }

        NotificationSaveVo vo = new NotificationSaveVo();
        vo.setOk(true);
        vo.setId(dto.getNotificationId());
        return vo;
    }
}
